html_named_entities/build/main.rs

```diff
@@ -0,0 +1,373 @@
+
+        if first.is_terminal != second.is_terminal {
+            return false;
+        }
+
+        // First check if exactly the same edges are present
+        if first
+            .edges
+            .iter()
+            .zip(second.edges.iter())
+            .any(|(a, b)| a.is_some() != b.is_some())
+        {
+            return false;
+        }
+
+        // The check whether the subtrees at each edge are equal.
+        first
+            .edges
+            .iter()
+            .zip(second.edges.iter())
+            .filter_map(|(first_edge, second_edge)| first_edge.zip(*second_edge))
+            .all(|(first_edge, second_edge)| self.are_subtrees_equal(first_edge, second_edge))
+    }
+
+    fn compute_numbers_for(&mut self, index: usize) {
+        if self.nodes[index].num_nodes != 0 {
+            // We already computed the unique value for this node
+            return;
+        }
+
+        self.nodes[index].num_nodes += self.nodes[index].is_terminal as usize;
+        for edge in self.nodes[index].edges {
+            let Some(edge) = edge else {
+                continue;
+            };
+
+            self.compute_numbers_for(edge);
+            self.nodes[index].num_nodes += self.nodes[edge].num_nodes;
+        }
+    }
+
+    /// Computes all numbers needed for minimal perfect hashing
+    fn compute_numbers(&mut self) {
+        self.compute_numbers_for(0);
+    }
+
+    /// Returns the perfect hash value for the input, or `None` if
+    /// the input was not in the input set.
+    ///
+    /// Hashing is done by computing a prefix sum over `Node::num_nodes` at
+    /// every step of the traversal.
+    fn get_unique_index(&self, input: &str) -> Option<usize> {
+        assert!(input.is_ascii());
+
+        let mut index = 0;
+        let mut current = &self.nodes[0];
+        for code_point in input.as_bytes() {
+            let next_node = current.edges[*code_point as usize]?;
+
+            for edge in &current.edges[..*code_point as usize] {
+                let Some(edge) = edge else {
+                    continue;
+                };
+                index += self.nodes[*edge].num_nodes;
+            }
+
+            current = &self.nodes[next_node];
+            if current.is_terminal {
+                index += 1;
+            }
+        }
+
+        debug_assert!(
+            current.is_terminal,
+            "Traversing {input:?} did not end at a terminal node"
+        );
+        Some(index)
+    }
+}
+
+impl Node {
+    fn edges(&self) -> impl Iterator<Item = usize> + use<'_> {
+        self.edges.iter().filter_map(|edge| *edge)
+    }
+}
+
+#[derive(Deserialize)]
+struct NamedEntity {
+    codepoints: Vec<u32>,
+    #[allow(dead_code)]
+    characters: String,
+}
+
+fn main() {
+    println!("cargo:rerun-if-changed=build/");
+
+    // Parse the list of named entities from https://html.spec.whatwg.org/entities.json
+    let input_file = BufReader::new(File::open("build/entities.json").unwrap());
```

Comment:
Consider converting the JSON file into a Rust file. Then you wouldn't need to depend on `serde` in the build.rs.